Lbcheng

lbcheng888 opened the pull request:
- Documented all ARC/ORC deterministic-release work, including IR generation changes (nimTraceClosure, mandatory attachedTrace),
    runtime support (TNimTypeV2, cyclebreaker.releaseGraph/thinout, lightweight nimTraceRef* stub), and async lifecycle fixes
    (callSoon, consolidated leak tests, new stress suites).
  - Published the verification matrix that was used to vet the new behavior (ARC/ORC runs, threaded scenarios, leak counters).

lbcheng888 committed:
Add ARC/ORC regression for async callSoon closures

lbcheng888 committed:
Merge pull request #1 from lbcheng888/codex/fix-async-memory-leak-in-arc/orc

Add ARC/ORC regression for async callSoon closures

lbcheng888 committed:
Ensure async callbacks release closure environments

lbcheng888 committed:
Merge branch 'devel' into codex/fix-async-memory-leak-in-arc/orc-n2w680

Commit:
update fix memory leak

Commit:
- Documented all ARC/ORC deterministic-release work, including IR generation changes (nimTraceClosure, mandatory attachedTrace),
    runtime support (TNimTypeV2, cyclebreaker.releaseGraph/thinout, lightweight nimTraceRef* stub), and async lifecycle fixes
    (callSoon, consolidated leak tests, new stress suites).
  - Published the verification matrix that was used to vet the new behavior (ARC/ORC runs, threaded scenarios, leak counters).

Araq commented:
Most interesting. Can you give us more documentation please?

Commit:
try to fix CI

Commit:
add orc_arc_fix.md

Commit:
fix  NIM_TESTAMENT_REMOTE_NETWORKING=1 ./koch tests cat nimble-packages important_packages

lbcheng888 commented:
sure, [b395500](https://github.com/nim-lang/Nim/pull/25225/commits/b395500efa34813777379ae0496ccd8669689099)

Araq commented on lib/system/orc.nim:
```diff
@@ -140,15 +140,25 @@ proc nimTraceRefDyn(q: pointer; env: pointer) {.compilerRtl, inl.} =
 
 proc unregisterCycle(s: Cell) =
   # swap with the last element. O(1)
+  if s.rootIdx == 0:
```
Seems fishy. Why can this happen now that we support thinOut?

lbcheng888 replied:
Thanks for reviewing and comment.

lbcheng888 replied:
`thinOut` only guarantees that the roots array is compact **at the moment it runs**.
    During destructors or partial collections `roots` can be mutated again and the
    cached `rootIdx` becomes stale (it can even drop to zero although the entry is
    still present). With thinOut alone we can still reach `unregisterCycle` with an
    invalid index. The fix recompacts `roots` on demand and rescans for the cell so
    the removal is always correct.

Commit:
comply to thinOut solution

Commit:
orc: fix stale rootIdx after thin-out

Destructors and partial collects can reshuffle the ORC roots array, leaving
entries with rootIdx <= 0 even though the cell remains tracked. The old
unregisterCycle treated that as already-unlinked and leaked the root.

Add an internal compactRoots helper that recomputes indices, let
unregisterCycle revalidate its slot (falling back to a scan if thin-out moved
the cell), and only then swap it with the tail.

Tests: testament/testament --nim:bin/nim c arc "--mm:arc"
       testament/testament --nim:bin/nim c arc "--mm:orc"
       nim c --gc:orc -d:nimThinout -r tests/arc/thamming_thinout.nim

arnetheduck commented:
[chronos](https://github.com/status-im/nim-chronos/pull/243) already achieves similar memory release properties without the additional complexity of runtime tracing simply by improving the implementation of the dispatch pipeline - are there cases that this approach covers, which the chronos approach does not?
